[JIT] add support for lu_unpack

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33818 [JIT] Add support for list()
* #33783 [JIT] add support for torch.norm
* #33737 [JIT] add support for torch.cdist
* **#33736 [JIT] add support for lu_unpack**
* #33724 [JIT] add support for torch.lu to torchscript
* #33504 [JIT] fix resolving of functions in torch/functional. fix compilation of torch.stft

Differential Revision: [D20121914](https://our.internmc.facebook.com/intern/diff/D20121914)

Commit:
[JIT] add support for lu_unpack

[ghstack-poisoned]

vishwakftw commented on test/test_jit.py:
```diff
@@ -10204,7 +10204,15 @@ def lu_infos(x):
             return torch.lu(x, get_infos=True)
 
-        self.checkScript(lu_infos, (torch.randn(2, 3, 3),))
+        self.checkScript(lu, (torch.randn(2, 3, 3),))
```
I believe this got changed back erroneously.

vishwakftw commented on torch/functional.py:
```diff
@@ -87,8 +86,28 @@ def split(tensor, split_size_or_sections, dim=0):
     return tensor.split(split_size_or_sections, dim)
 
+# equivalent to itertools.product(indices)
+def indices_product(indices):
```
Is there a way to bypass this when JIT isn't on?

eellison replied:
it's possible, but why exactly ? there should be a high bar for introducing divergent behavior for JIT & Python. 
This function produces the exact same values as `product`. When I ran 
```
import time
now = time.perf_counter()
for i in range(100):
    for shape in ((3, 3), (5, 3, 3), (7, 3, 5, 5), (7, 5, 3, 3, 3)):
    a = torch.randn(*shape, dtype=dtype, device=device)
    a_lu, p = torch.lu(a, pivot=pivot)
    p_ref, l_ref, u_ref = torch.lu_unpack(a_lu, p)
print(time.perf_counter() - now)
```
I consistently got 6.5 seconds for this implementation, and 7 seconds for the itertools.product implementation. So it also seems to be (somewhat) faster.

vishwakftw replied:
I thought there might be a performance issue, but evidently not. Thank you for clarifying.

driazati commented on torch/functional.py:
```diff
@@ -87,8 +86,28 @@ def split(tensor, split_size_or_sections, dim=0):
+# equivalent to itertools.product(indices)
+def indices_product(indices):
+    # type: (List[int]) -> (List[List[int]])
+    empty_list : List[int] = []
```
We still have to support python 2 so I think you have to use `torch.jit.annotate`

driazati commented on torch/functional.py:
```diff
@@ -170,14 +189,20 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+                for k, j in enumerate(index_tensor_with_indices_list(LU_pivots_zero_idx, idx)):
                     final_order[k], final_order[j] = final_order[j], final_order[k]
-                P[idx] = P[idx].index_select(1, torch.as_tensor(final_order, device=LU_pivots.device))
+                # TODO: remove index_tensor_with_indices_list when TorchScript supports indexing Tensor with list
```
Can you file an issue for this and link it in the comment?

eellison replied:
when we had similar issues we just closed them - https://github.com/pytorch/pytorch/issues/32246, https://github.com/pytorch/pytorch/issues/27493

driazati commented on torch/functional.py:
```diff
@@ -170,14 +189,20 @@ def lu_unpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
-                for k, j in enumerate(LU_pivots_zero_idx[idx]):
+
+            # TODO: rewrite when TorchScript supports product and map as
+            # product(*map(lambda x: list(range(x)), shape[:-2]))
```
Same here about issue link

driazati reviewed (approved):
looks good after a couple small fixes

Commit:
Update on "[JIT] add support for lu_unpack"




[ghstack-poisoned]

Commit:
Update on "[JIT] add support for lu_unpack"




[ghstack-poisoned]

Commit:
Update on "[JIT] add support for lu_unpack"


Differential Revision: [D20121914](https://our.internmc.facebook.com/intern/diff/D20121914)

[ghstack-poisoned]

Commit:
Update on "[JIT] add support for lu_unpack"


Differential Revision: [D20121914](https://our.internmc.facebook.com/intern/diff/D20121914)

[ghstack-poisoned]
